File: graph/488-checkSCC.py
from collections import defaultdict

def checkSCC(n,edges):
    
    def check(cur,G):
        seen=set()
        
        def dfs(cur):
            if cur in seen:
                return
            seen.add(cur)
            for i in G:   #G[1]={2,3,4}
                dfs(i)
        dfs(cur)
        return len(seen)==n
    G=defaultdict(list)
    for a,b in edges:
        G[a].append(b) 
    ReversedG=defaultdict(list)
    for a,b in edges:
        ReversedG[b]=a
    
    return check(0,G) and check(0,ReversedG)          

n=2
roads=[[1,0],[0,1]]

print(checkSCC(n,roads))

File: graph/488-checkSCC.cpp
#include<vector>
#include<set>
#include<map>
#include<iostream>

using VEC=std::vector<int>;
using MATRIX=std::vector<std::vector<int>>;
using SET=std::set<int>;
using GRAPH=std::map<int,VEC>;


void dfs(int cur,const GRAPH& G,SET& s)
{
    if(s.find(cur)!=s.end())
        return ;
    s.insert(cur);
    for(auto &[k,v]:G)
    {
        dfs(k,G,s);
    }
}
bool check(int n,int cur,const GRAPH& G)
{
    SET seen{};
    dfs(cur,G,seen);
    return n==(int)(seen.size());
}
bool checkSCC(MATRIX& m,int n)
{
    GRAPH G1{},G2{};
    for(auto &v:m)
    {
        G1[v[0]].push_back(v[1]);
        G2[v[1]].push_back(v[0]);
    }
        return check(n,0,G1)&&check(n,0,G2);

}

int main()
{
    int n=2;
    MATRIX m{{0,1},{1,0}};
    std::cout<<checkSCC(m,2);
}